source/mir/math/stat.d

```diff
@@ -424,6 +426,241 @@ unittest
+version(mir_test_topN)
+@safe pure nothrow
+unittest {
+    import mir.ndslice.slice: sliced;
+
+    auto x0 = [9.0, 1, 0, 2, 3, 4, 6, 8, 7, 10, 5].sliced;
+    assert(x0.median == 5);
+
+    auto x1 = [9.0, 1, 0, 2, 3, 4, 6, 8, 7, 10].sliced;
+    assert(x1.median == 5);
+}
+
+/// Median of matrix
+version(mir_test_topN)
+@safe pure
+unittest {
+    import mir.ndslice.fuse: fuse;
+
+    auto x0 = [
+        [9.0, 1, 0, 2,  3], 
+        [4.0, 6, 8, 7, 10]
+    ].fuse;
+
+    assert(x0.median == 5);
+}
+
+/// Row median of matrix
+version(mir_test_topN)
+@safe pure
+unittest
+{
+    import mir.ndslice.fuse: fuse;
+    import mir.ndslice.slice: sliced;
+    import mir.ndslice.topology: alongDim, byDim, map;
+    import mir.algorithm.iteration: all;
+    import mir.math.common: approxEqual;
+
+    auto x = [
+        [0.0, 1.0, 1.5, 2.0, 3.5, 4.25], 
+        [2.0, 7.5, 5.0, 1.0, 1.5, 0.0]
+    ].fuse;
+
+    auto result = [1.75, 1.75].sliced;
+
+    // Use byDim or alongDim with map to compute median of row/column.
+    assert(x.byDim!0.map!median.all!approxEqual(result));
+    assert(x.alongDim!1.map!median.all!approxEqual(result));
+}
+
+/// Can allow original slice to be modified or set output type
+version(mir_test_topN)
+@safe pure nothrow
+unittest {
+    import mir.ndslice.slice: sliced;
+
+    auto x0 = [9.0, 1, 0, 2, 3, 4, 6, 8, 7, 10, 5].sliced;
+    assert(x0.median!(Yes.allowModify) == 5);
+    
+    auto x1 = [9, 1, 0, 2, 3, 4, 6, 8, 7, 10].sliced;
+    assert(x1.median!(float, Yes.allowModify) == 5);
+}
+
+/++
+For integral slices, pass output type as template parameter to ensure output
+type is correct
++/
+version(mir_test_topN)
+@safe pure nothrow
+unittest {
+    import mir.ndslice.slice: sliced;
+
+    auto x = [9, 1, 0, 2, 3, 4, 6, 8, 7, 10].sliced;
+    assert(x.median!float == 5);
+}
+
+version(mir_test)
+@safe pure nothrow
+unittest {
+    import mir.ndslice.slice: sliced;
+    import mir.math.common: approxEqual;
+
+    auto x0 = [9.0, 1, 0, 2, 3].sliced;
+    assert(x0.median.approxEqual(2));
+
+    auto x1 = [9.0, 1, 0, 2].sliced;
+    assert(x1.median.approxEqual(1.5));
+    
+    auto x2 = [9.0, 0, 1].sliced;
+    assert(x2.median.approxEqual(1));
+    
+    auto x3 = [1.0, 0].sliced;
+    assert(x3.median.approxEqual(0.5));
+    
+    auto x4 = [1.0].sliced;
+    assert(x4.median.approxEqual(1));
+}
+
+private pure @safe nothrow @nogc
+F medianImpl(F, Iterator, SliceKind kind)(Slice!(Iterator, 1, kind) slice)
+{
+    import mir.ndslice.sorting: topN;
+
+    size_t len = slice.length;
+    if (len > 5) {
+        size_t half_n = len / 2;
+        topN(slice, half_n);
+        if (len % 2 == 1) {
+            return cast(F) slice[half_n];
+        } else {
+            //move largest value in first half of slice to half_n - 1
+            topN(slice[0 .. half_n], half_n - 1);
+            return mean!F(slice[(half_n - 1) .. (half_n + 1)]);
+        }
+    } else {
+        return smallMedianImpl!(F)(slice);
+    }
+}
```

Comment:
The algorithm looks weird. Needs many more tests. It is good to generate a set of samples and add them to unit tests.

`partitionAt` and `topN` are similar, but they are not the same.

`partitionAt`:  `max(a[0 .. i]) <= a[i] <= min(a[i+1 .. $])`
`topN`:  `max(a[0 .. i]) <= min(a[i+1 .. $])`

Also, the `mean` should have exactly one (or maybe zero) calls of `topN ` /  `partitionAt`.
Instead of adding a second one call `topN ` /  `partitionAt` it is better to do  2 (for `topN`) / 1 (for `partitionAt`) calls of min or max handles.  `partitionAt` is better for mean because it requires only on of min or max handles.

If `topN` uses `pivotPartiton` is already or almost `partitionAt`.

Reply:
`pivotPartition` does the same as what you describe as `pivotAt`, I believe. Currently, `topN` uses `pivotPartition`, except if n == 0 or (n + 1) == slice.length. There are specializations in there that handle those cases (in a way that may be more efficient than `pivotPartition`. As a result, the call `topN(slice[0 .. half_n], half_n - 1);` puts the largest value of slice[0 ..half_n] at slice[half_n - 1]. Similarly, `topN(slice[0 .. half_n], 0);` would have put the smallest value at slice[0]. I think you're right though that I would need to two calls. It may be possible that the right-hand side has a value equal to slice[half_n] and I need to check for that.

Also, do you mean `median` in some of that rather than `mean`? The call to `mean` there is used just because there are two values to be averaged and I thought it made more sense to re-use what is already in there if I could.

Reply:
Also, the documentation for `topN` seems consistent with what you have for `pivotAt`, at least as far as I can tell. 

The documentation has `r[0]` to `r[nth]`, so that would be `a[0..(i+1)]` in your notation. The next lines in the documentation are equivalent to
`not(max(a[0 .. (i + 1)]) > a[i])`
and 
`not(a[i] > min(a[i .. $]))`
Together, they should be equivalent to  
`topN: max(a[0 .. i]) <= a[i] <= min(a[i+1 .. $])`

Reply:
Yes, I mean median, not mean

Reply:
Do you mean our docs or Phobos or both? Does the Phobos topN do the same? I thoughtit isn’t. If so, it then topN is the incorrect naming. topN should just split into two parts, and we shouldn’t expect more from this name.

Reply:
yeah, I should have write pivotAt

Reply:
I can change the name of `topN` to `partitionAt` or `pivotAt`, no real preference on my part. Nevertheless, the `topN` function in mir behaves similarly to the phobos `topN` function. The two main differences are in what it returns and that the phobos function uses a deterministic algorithm that can perform better for large arrays (mir's just allows a pivotFunction to be called, but they should be the same for smaller arrays with `setPivotAt`). However, both should have the effect of enforcing `max(a[0 .. i]) <= a[i] <= min(a[i+1 .. $])` for our case. I effectively check for this in the unittests that call `checkTopNAll`. This has implications for the `median` algorithm. 

Suppose if we have some initial `a`, where `len = a.length` and `half_len=len / 2` and `len%2 == 0` (so len is even). The simple way to calculate `median` in this case is to sort `a` and return `mean(a[half_len - 1], a[half_len])`. If, instead, we call `partitionAt(a, half_n)`, then there are four possibilities:

1. `max(a[0 .. half_n]) < a[half_n],  a[half_n] < min(a[half_n+1 .. $])`.
2. `max(a[0 .. half_n]) < a[half_n],  a[half_n] = min(a[half_n+1 .. $])`.
3. `max(a[0 .. half_n]) = a[half_n],  a[half_n] < min(a[half_n+1 .. $])`
4. `max(a[0 .. half_n]) = a[half_n],  a[half_n] = min(a[half_n+1 .. $])`

We can think of these cases as `1` means that `a[half_n]` appears once. `2` and `3` mean that it appears two or more times. In `2`  the extra `a[half_n]` only appear on the right side and in `3` it only appears on the left side. `4` means that it is split on the left and right sides (for a total of three or more). 

In `1`, `median` equals `mean(max(a[0 .. half_n])), a[half_n])` because it is obviously equivalent to what you would get it if you sort. For `2`, everything left of `half_n` is less than `a[half_n]`. While there is a value to the right of `half_n` equal to `a[half_n]`, it does not matter for the purposes of calculating the `median`. The formula is the same as `1`, `mean(max(a[0 .. half_n])), a[half_n])`. `3` and `4` are both similar as well. For `3`, everything to the right of `half_n` is greater than `a[half_n]`, but there are some values to the left of `half_n` that are equal to `a[half_n]`. This means that the `median` equals `mean(a[half_n], a[half_n])=a[half_n]`. In `4`, the `median` equals `a[half_n]` because a[half_n] is on both the left and right. This is equivalent to `mean(max(a[0 .. half_n])), a[half_n])`.

Thus, so long as we have a function that does `max(a[0 .. i]) <= a[i] <= min(a[i+1 .. $])`, we can calculate the `median` for arrays of even length by calling it first for the whole `a` and then by calling it for for `a[0 .. i]` to get the maximum of the left side.